### ISSUE TITLE
Twrp recovery won't work on OnePlus 3 after 2.5.1 update

### ISSUE THREAD

**Issue by nathanatgit** [TeamWin/Team-Win-Recovery-Project#759]
The twrp-3.0.2-1 recovery failed to start, even run again "fastboot flash recovery twrp-3.0.2-1-oneplus3.img" in fastboot mode. Occurs after the official h2os update to 2.5.1.
btw, the official h2os 2.5.1 update zip file seems to have changed that it can't be directly flashed via twrp. But I used to flash an old 2.5.0 update via twrp and it works fine.
I have to change back to the original recovery to run 2.5.1 update. However when the update was finished, I flashed twrp again in fastboot mode and then reboot to recovery, the phone failed to boot and ended up in black screen.

**Comment by cfaber** [comment]
what do you mean by this? I'm running twrp 3.0.2-1 with h2Os 2.5.1 without issue. The only problem I'm seeing is https://github.com/TeamWin/Team-Win-Recovery-Project/issues/760

**Comment by nathanatgit** [comment]
Saw it and problem solved by removing scrip `getprop("ro.build.product") == "OnePlus3" || abort("This package is for "OnePlus3" devices; this is a "" + getprop("ro.build.product") + "".");` in official zip file. Thanks!

**Comment by nathanatgit** [comment]
but I can‘t use the twrp recovery after the update. I flashed the 3.0.2-1 image again in my phone and choose to boot in to recovery, I see oneplus 3 logo and then a black screen, nothing appears later. I can boot into system without a problem but I just can't boot into twrp recovery.
now i have to choose some version of twrp recovery modified by forum members as a temporary solution.

**Comment by cfaber** [comment]
Are you trying to run with oem lock enabled by chance?

**Comment by nathanatgit** [comment]
No, I'm sure my phone is still unlocked.there is unlock warning screen at start up.

One strange thing I forget to say:

After the 2.5.1 update, the h2os allows user to turn on advanced reboot menu in developers option .I can choose to reboot into recovery in reboot through long-press power button. 

However no matter how many times I run the command in fastboot mode to flash twrp image into my phone, when I choose to reboot to recovery in the **Advanced reboot menu**, the phone ended up booting  in to the official recovery. 

I'm pretty pretty sure that my phone stays unlocked and I'm sure I flashed twrp img successfully in to the recovery sector.
This only happens if I choose to reboot into recovery through advanced boot menu. When I choose to boot into recovery at the start up unlock warning screen, the phone can boot into twrp recovery but ended in a black screen.the twrp seems be stuck after the oneplus  logo disspeared.
Same thing DIDN'T happen if I flash a forum modified version of twrp img, and phone can boot into twrp recovery successfully, no matter whether choosing to boot into recovery in at the start up unlock warning screen OR after logging into system and rebooting through a long press to call Advanced reboot menu and choose 'reboot into recovery' in the menu.

Hope that it would help.

**Comment by cfaber** [comment]
How are you flashing recovery, can you post exact commands?

**Comment by nathanatgit** [comment]
fastboot flash recovery twrp-3.0.2-1-oneplus3.img

**Comment by cfaber** [comment]
This should work, maybe try and rename to recovery.img (not sure if
fastboot wants or cares about the image name itself)

**Comment by nathanatgit** [comment]
Tried, because I flashed it many many times so I rename it into t.img for convenient. Same problem happens.

**Comment by cfaber** [comment]
Hm... Are you running a recent version of fastboot?

**Comment by nathanatgit** [comment]
![fotor_147456712626187](https://cloud.githubusercontent.com/assets/20399271/18759921/6ad97776-80c4-11e6-98e5-59ed4edbe46b.jpg)
Yes, the fastboot works fine.
I posted a screenshot when you search"twrp 2.5.1" in the official Chinese forum.
Clearly many people are facing the same problem.

**Comment by nathanatgit** [comment]
Here is the modified twrp recovery I'm using as a temporary solution
![fotor_147456823735225](https://cloud.githubusercontent.com/assets/20399271/18760558/3809602e-80c7-11e6-9c45-ea771f40ef5a.jpg)
.

**Comment by lochuan** [comment]
Yep, After update to Hydrogen OS 2.5.1, the twrp 3.0.2-1 does't work. It just show you a black screen when you wanna jump into recovery mode

**Comment by jfeise** [comment]
Same issue with the official CM-14.1 nightlies for the OP3.
At http://forum.xda-developers.com/oneplus-3/how-to/rom-community-build-3-5-5-t3490939 there is a link to a modified TWRP that fixes the problem.

**Comment by JayceJocelyn** [comment]
@jfeise Thank you so much! Everything is just working fine now. 👍 💯

**Comment by hellow554** [comment]
This problem is still present. Why can't it be fixed upstream?!

**Comment by jcadduono** [comment]
happy now?
(you guys should have told me earlier :))

**Comment by hellow554** [comment]
Do you have a commit hash? (on which branch?)

**Comment by jcadduono** [comment]
android-6.0 the link is on the device download page on TWRP.me

**Comment by hellow554** [comment]
https://dl.twrp.me/oneplus3/
3.0.3 works for me :) thanks

**Comment by jfeise** [comment]
3.0.3 does not decrypt a data partition encrypted with password.

**Comment by jcadduono** [comment]
use 3.0.4 build from https://forum.xda-developers.com/oneplus-3/development/recovery-official-twrp-oneplus-3-3t-t3543391 for now which adds ext4 data decryption
if you still have a problem, you won't get any help without providing dmesg and recovery.log

**Comment by jfeise** [comment]
3.0.4 works. Thanks for the quick answer.

**Comment by jfeise** [comment]
While the decryption works with 3.0.4, it has a bunch of other problems, like reboot not working. I had to go back to the unofficial TWRP 3.0.2-22 that I mentioned earlier.

**Comment by jcadduono** [comment]
no logs no proof no care

**Comment by jfeise** [comment]
Can't reproduce it right now, unfortunately. I had installed it through the app, maybe that caused the issues.

**Comment by jfeise** [comment]
Was able to reproduce it now:
Factory reset, then flash LineageOS nightly, then reboot, result: recovery shows up again, without actual reboot.
recovery log attached. df shows /sdcard mounted twice:

> ~# df
Filesystem           1K-blocks      Used Available Use% Mounted on
tmpfs                  2921832        24   2921808   0% /dev
tmpfs                  2921832      5240   2916592   0% /tmp
/dev/block/dm-0       55224864  23956108  31252372  43% /sdcard
/dev/block/sda3         253920      8892    239788   4% /cache
/dev/block/dm-0       55224864  23956108  31252372  43% /data
/dev/block/dm-0       55224864  23956108  31252372  43% /sdcard
>
[recovery.txt](https://github.com/TeamWin/Team-Win-Recovery-Project/files/734612/recovery.txt)

After that, any reboot selection in the Reboot menu just goes back to recovery, with each "reboot" just increasing the number of times /sdcard is mounted.
The only way to get out of that is to use "adb reboot". That does a proper reboot.

**Comment by jcadduono** [comment]
k this is a side effect of the decryption multi mount patch will need to look into it..
